src/repo/mod.rs

```diff
@@ -111,10 +114,67 @@ pub trait DataStore: Debug + Send + Sync + Unpin + 'static {
     async fn wipe(&self);
 }
 
+#[async_trait]
+pub trait PinStore: Debug + Send + Sync + Unpin + 'static {
+    async fn is_pinned(&self, block: &Cid) -> Result<bool, Error>;
+
+    async fn insert_pin(&self, target: &Cid, kind: PinKind<&'_ Cid>) -> Result<(), Error>;
+
+    async fn remove_pin(&self, target: &Cid, kind: PinKind<&'_ Cid>) -> Result<(), Error>;
+
+    async fn list(
+        &self,
+        mode: Option<PinMode>,
+    ) -> futures::stream::BoxStream<'static, Result<(Cid, PinMode), Error>>;
+
+    // here we should have resolved ids
+    // go-ipfs: doesnt start fetching the paths
+    // js-ipfs: starts fetching paths
+    // FIXME: there should probably be an additional Result<$inner, Error> here; the per pin error
+    // is serde OR cid::Error.
+    /// Returns error if any of the ids isn't pinned in the required type, otherwise returns
+    /// the pin details if all of the cids are pinned in one way or the another.
+    async fn query(
+        &self,
+        ids: Vec<Cid>,
+        requirement: Option<PinMode>,
+    ) -> Result<Vec<(Cid, PinKind<Cid>)>, Error>;
+}
+
 #[derive(Clone, Copy, Debug)]
 pub enum Column {
     Ipns,
-    Pin,
+}
+
+/// `PinMode` is the description of pin type for quering purposes.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum PinMode {
+    Indirect,
+    Direct,
+    Recursive,
+}
+
+impl<'a, B: Borrow<Cid>> PartialEq<&'a PinMode> for PinKind<B> {
```

Comment:
does `B` need to be specified here? I'd expect the compiler to be able to implement this for any `B`, but I might be giving it a bit too much credit

Reply:
TBH I think we always need to repeat all kinds of bounds and I haven't at least seen any news that implied bounds was a feature, perhaps in some lifetime cases...?

If you meant this as `impl<'a, B> PartialEq<&'a PinMode> for PinKind<B> { }` but my inner rustc is already telling me this type is not valid `PinKind` because the bound is on the `enum`. I guess I *could* remove the bound from the enum decl but I kinda like it would lose the readability, if you couldn't deduce what the `B` is just by looking at the types.

Reply:
Did not test this, only removed the ampersand.